src/raftstore/store/worker/apply.rs

```diff
@@ -914,39 +984,46 @@ impl ApplyDelegate {
-            AdminCmdType::ComputeHash => self.exec_compute_hash(ctx, request),
-            AdminCmdType::VerifyHash => self.exec_verify_hash(ctx, request),
+        box match cmd_type {
+            AdminCmdType::ChangePeer => self.exec_change_peer(ctx, &request),
+            AdminCmdType::Split => self.exec_split(ctx, &request),
+            AdminCmdType::CompactLog => self.exec_compact_log(ctx, &request),
+            AdminCmdType::TransferLeader => {
+                box future::err((self, ctx, box_err!("transfer leader won't exec")))
+            }
+            AdminCmdType::ComputeHash => self.exec_compute_hash(ctx, &request),
+            AdminCmdType::VerifyHash => self.exec_verify_hash(ctx, &request),
             // TODO: is it backward compatible to add new cmd_type?
-            AdminCmdType::PrepareMerge => self.exec_prepare_merge(ctx, request),
-            AdminCmdType::CommitMerge => self.exec_commit_merge(ctx, request),
-            AdminCmdType::RollbackMerge => self.exec_rollback_merge(ctx, request),
-            AdminCmdType::InvalidAdmin => Err(box_err!("unsupported admin command type")),
-        }?;
-        response.set_cmd_type(cmd_type);
-
-        let mut resp = RaftCmdResponse::new();
-        let uuid = ctx.exec_ctx
-            .as_ref()
-            .unwrap()
-            .req
-            .get_header()
-            .get_uuid()
-            .to_vec();
-        resp.mut_header().set_uuid(uuid);
-        resp.set_admin_response(response);
-        Ok((resp, exec_result))
+            AdminCmdType::PrepareMerge => self.exec_prepare_merge(ctx, &request),
+            AdminCmdType::CommitMerge => self.exec_commit_merge(ctx, &request),
+            AdminCmdType::RollbackMerge => self.exec_rollback_merge(ctx, &request),
+            AdminCmdType::InvalidAdmin => {
+                box future::err((self, ctx, box_err!("unsupported admin command type")))
+            }
+        }.then(move |res| match res {
+            Ok((delegate, ctx, mut response, exec_result)) => {
+                response.set_cmd_type(cmd_type);
+                let mut resp = RaftCmdResponse::new();
+                let uuid = ctx.exec_ctx
+                    .as_ref()
+                    .unwrap()
+                    .req
+                    .get_header()
+                    .get_uuid()
+                    .to_vec();
+                resp.mut_header().set_uuid(uuid);
+                resp.set_admin_response(response);
+                future::ok((delegate, ctx, resp, exec_result))
+            }
+            Err((delegate, ctx, e)) => future::err((delegate, ctx, e)),
+        })
     }
 
     fn exec_change_peer(
-        &mut self,
-        ctx: &mut ApplyContext,
+        mut self,
+        mut ctx: ApplyContext,
         request: &AdminRequest,
-    ) -> Result<(AdminResponse, Option<ExecResult>)> {
+    ) -> Box<Future<Item = AdminCmdFutureItem, Error = FutureError> + Send> {
```

Comment:
I don't see why it is returning a future. Seems that it uses only normal results. BTW, you can return future::result, which is the same type for all exit paths, so that you don't need to Box<> it.